### PR TITLE
#167484558 Users should be able to follow each other

### PR DESCRIPTION
**Pull request by raymond42**
### What does this PR do?
- Users should be able to follow each other
### Description of Task to be completed?
- Create button for following and unfollowing
- Connect the follow and unfollow button to the backend
- Put the follow and unfollow button on a single article fetch
### How should this be manually tested?
- Clone this branch
- `npm install` to install all dependencies
- `npm start` to run the server
- `Signup or login to follow or unfollow an author`]
- View a single article
- click follow or unfollow button to follow or unfollow the author
### Any background context you want to provide?
- N/A
### What are the relevant pivotal tracker stories?
- [Users should be able to follow each other](https://www.pivotaltracker.com/story/show/167484558)

### CODE REVIEW

**Review by placideirandora — CHANGES_REQUESTED** [review]
When I try to follow an author, I receive the error: jwt malformed.

It will be fixed

**Review by malaba6 — CHANGES_REQUESTED** [review]
could you fix the failing tests on Travis first before we can review your test?

**Review by Jkadhuwa — CHANGES_REQUESTED** [review]
The functionality works as intended, but could you please improve the responsiveness on some of the devices like a screen of 1280px * 800px.

**Review by Jkadhuwa — APPROVED** [review]

**Comment by raymond42** [comment]
> The functionality works as intended, but could you please improve the responsiveness on some of the devices like a screen of 1280px * 800px.

done. Thanks

**Comment by raymond42** [comment]
> could you fix the failing tests on Travis first before we can review your test?

done. Thanks

**Comment by raymond42** [comment]
> When I try to follow an author, I receive the error: jwt malformed.
> 
> It will be fixed

Implemented your feedback

**Comment by raymond42** [comment]
> > When I try to follow an author, I receive the error: jwt malformed.
> > It will be fixed
> 
> @placideirandora Fixed. Thanks

**Review by placideirandora — APPROVED** [review]

**Review by malaba6 — APPROVED** [review]

File: src/app/App.js [hunk 1]
```diff
@@ -29,13 +32,16 @@ function App() {
...
           <Route path="/login" component={Login} />
           <Route path="/signup" component={SignUp} />
           <Route path="/profile" component={Profile} />
           <Route path="/update-profile" component={UpdateProfile} />
           <Route path="/Signup" component={SignUp} />
-          <Route path="/articles/:slug" component={SingleArticle} />
-          <ProtectedRoutes path="/Create" component={CreateArticle} />
+          <Route path="/profiles/:userName/follow" component={FollowUnfollowComponent} />
+          <Route path="/profiles/:userName/unfollow" component={FollowUnfollowComponent} />
+          {/* // <Route path="/articles/:slug" component={SingleArticle} /> */}
```
Comment (bdushimi) [inline src/app/App.js]: @raymond42 Kindly remove such commented lines/codes

File: src/feature/articles/__tests__/CreateArticleAction.test.js [hunk 2]
```diff
@@ -42,6 +42,23 @@ describe('Create Article success', () => {
...
+      response: expected
+    });
+    const mockArticles = {
+      description: 'I knew that would happen soon but i kept it inside',
+      body: 'I knew that would happen soon but i kept it inside'
+    };
+    return store.dispatch(createArticle(mockArticles)).then(() => {
+      expect(store.getActions().length).toEqual(1);
+    });
+  });
```
Comment (bdushimi) [inline src/feature/articles/__tests__/CreateArticleAction.test.js:61]: @raymond42 any reason to why this PR for following contains changes in unrelated files i.e create article actions? This PR should contain only changes related to one functionality i.e. follow and unfollow a user

File: src/feature/articles/constants.js [hunk 3]
```diff
@@ -2,3 +2,10 @@ export const CREATE_ARTICLE_SUCCESS = 'CREATE_ARTICLE_SUCCESS';
 export const CREATE_ARTICLE_FAIL = 'CREATE_ARTICLE_FAIL';
 export const GET_ARTICLES_SUCCESS = 'GET_ARTICLES_SUCCESS';
 export const GET_ARTICLES_FAIL = 'GET_ARTICLES_FAIL';
+export const GET_SINGLE_ARTICLE_SUCCESS = 'GET_SINGLE_ARTICLE_SUCCESS';
+export const GET_SINGLE_ARTICLE_FAIL = 'GET_SINGLE_ARTICLE_FAIL';
+export const LIKE_ARTICLE_SUCCESS = 'GET_SINGLE_ARTICLE_SUCCESS';
+export const DISLIKE_ARTICLE_FAIL = 'GET_SINGLE_ARTICLE_FAIL';
+export const LIKE_ARTICLE_FAIL = 'GET_SINGLE_ARTICLE_SUCCESS';
+export const DISLIKE_ARTICLE_SUCCESS = 'GET_SINGLE_ARTICLE_FAIL';
```
Comment (bdushimi) [inline src/feature/articles/constants.js]: @raymond42 any reason to why this PR for following introduces changes which are not related to its functionality e.g LIKE_ARTICLE_SUCCESS, DISLIKE_ARTICLE_SUCCESS? This PR should contain only changes related to one functionality i.e. follow and unfollow a user.

File: src/feature/articles/createArticle/CreateArticleComponent.js [hunk 4]
```diff
@@ -8,7 +8,7 @@ import createArticle from './createArticleAction';
 import editorConfigs from '../../../app/config/ckeditorConfig';
 import 'react-toastify/dist/ReactToastify.css';
 import './createArticle.scss';
-import Home from '../../homePage/Home';
+// import Home from '../../homePage/Home';
```
Comment (bdushimi) [inline src/feature/articles/createArticle/CreateArticleComponent.js]: @raymond42 Kindly remove commented codes

File: src/feature/followUnfollow/FollowUnfollowComponent.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+const mapStateToProps = state => ({
+  following: state.followAuthor.following,
+  follow: state.followAuthor.follow,
+  loggedInUser: state.login
+});
+export default connect(
+  mapStateToProps,
+  {
+    clearFollowing, getFollowing, followAuthor, unfollowAuthor
+  }
```
Comment (bdushimi) [inline src/feature/followUnfollow/FollowUnfollowComponent.js]: @raymond42 while passing this object which contains functions/action creators works, I suggest you use mapDispatchProps function instead.

File: src/feature/followUnfollow/__tests__/followUnfollowAction.test.js [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+} from '../followUnfollowAction';
+import LocalStorage from '../../../__mocks__/localStorage';
+
+const middleware = [thunk];
+const mockStore = configureStore(middleware);
+const store = mockStore({});
+
+let storage;
+const token = 'Invalid token';
+describe('Follow Author Tests', () => {
```
Comment (bdushimi) [inline src/feature/followUnfollow/__tests__/followUnfollowAction.test.js]: @raymond42 the first argument in the describe block should be the name of the component under test., I think it should be 'FollowUnfollow ActionCreator'.

File: src/feature/followUnfollow/__tests__/followUnfollowAction.test.js [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+        status: 200,
+        response: {
+          data: {
+            message: 'Successfully rated this article'
+          }
+        }
+      });
+    });
+    return store.dispatch(followAuthor()).then(() => {
+      expect(store.getActions().length).toEqual(1);
```
Comment (bdushimi) [inline src/feature/followUnfollow/__tests__/followUnfollowAction.test.js]: @raymond42 while expecting that store.getActions().length equals to 1, a better and sure approach is to ascertain that the correct, a specific action has been dispatched. eg. store.getActions().toEqual(expectedAction). Kindly update where it is applicable in your tests.

File: src/feature/followUnfollow/__tests__/followUnfollowAction.test.js [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+    moxios.install(axios);
+    storage = window.localStorage.setItem('token', token);
+    window.localStorage = new LocalStorage();
+  });
+  afterEach(() => {
+    moxios.uninstall(axios);
+    store.clearActions();
+    window.localStorage = storage;
+  });
+  test('Follow Author Tests', () => {
```
Comment (bdushimi) [inline src/feature/followUnfollow/__tests__/followUnfollowAction.test.js]: @raymond42 Aren't we supposed to use 'describe' and 'it' blocks? Kindly update from test to it

File: src/feature/followUnfollow/__tests__/followUnfollowComponent.test.js [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+describe('Test FollowUnfollow Component...', () => {
+  it('Should render FollowUnfollow component', () => {
+    wrapper.setProps({ following: mockData.following });
+    expect(wrapper).toHaveLength(1);
+    wrapper.instance().UNSAFE_componentWillUnmount();
+  });
+  it('should update props', () => {
+    wrapper.setProps({ isAuthenticated: false });
+    expect(wrapper).toHaveLength(1);
```
Comment (bdushimi) [inline src/feature/followUnfollow/__tests__/followUnfollowComponent.test.js]: @raymond42 , which props is being ascertained that it got updated? The test description is not descriptive enough. It should indicate the props which are expected to be updated.

File: src/feature/followUnfollow/__tests__/followUnfollowComponent.test.js [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+const wrapper2 = renderFollowUnfollow({
+  loggedInUser: {
+    isAuthenticated: false,
+  }
+});
+
+describe('Test FollowUnfollow Component...', () => {
+  it('Should render FollowUnfollow component', () => {
+    wrapper.setProps({ following: mockData.following });
+    expect(wrapper).toHaveLength(1);
```
Comment (bdushimi) [inline src/feature/followUnfollow/__tests__/followUnfollowComponent.test.js:46]: @raymond42 , how's this test ascertain that the followUnfollow component is rendered? Just ascertain whether the wrapper could find some classes or html elements.

File: src/feature/followUnfollow/__tests__/followUnfollowComponent.test.js [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+  });
+  it('should update props', () => {
+    wrapper.setProps({ isAuthenticated: false });
+    expect(wrapper).toHaveLength(1);
+  });
+
+  it('should update props', () => {
+    wrapper.setProps({ isAuthenticated: true });
+    expect(wrapper).toHaveLength(1);
+  });
```
Comment (bdushimi) [inline src/feature/followUnfollow/__tests__/followUnfollowComponent.test.js]: @raymond42 , which props is being ascertained that it got updated? The test description is not descriptive enough. It should indicate the props which are expected to be updated.

File: src/feature/followUnfollow/__tests__/followUnfollowComponent.test.js [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+
+  test('Click Follow button...', () => {
+    const FollowButton = wrapper.find('.button');
+    FollowButton.simulate('click');
+  });
+  it('Click Follow button...', () => {
+    const FollowButton = wrapper2.find('.button');
+    FollowButton.simulate('click');
+  });
+});
```
Comment (bdushimi) [inline src/feature/followUnfollow/__tests__/followUnfollowComponent.test.js:71]: @raymond42 , what's the difference between the above 2 tests? The click event got simulated and nothing else! There're no assertions. You should expect some click handlers to have been got called.

File: src/feature/followUnfollow/followUnfollowAction.js [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+export const clearFollowing = () => (dispatch) => dispatch({ type: CLEAR_FOLLOW });
+
+export const getFollowing = () => async (dispatch) => {
+  const { token } = localStorage;
+  const axiosConfig = {
+    headers: {
+      'Content-Type': 'application/json',
+      Authorization: token
+    }
+  };
```
Comment (bdushimi) [inline src/feature/followUnfollow/followUnfollowAction.js]: @raymond42 wouldn't be better if such axiosConfiguration is centralized i.e. put in a file and get imported wherever and whenever one needs them?

**Review by bdushimi — CHANGES_REQUESTED** [review]

File: src/app/App.js (hunk 1 above)
Comment (raymond42) [reply src/app/App.js]: done

File: src/feature/articles/createArticle/CreateArticleComponent.js (hunk 4 above)
Comment (raymond42) [reply src/feature/articles/createArticle/CreateArticleComponent.js]: done

File: src/feature/followUnfollow/FollowUnfollowComponent.js (hunk 5 above)
Comment (raymond42) [reply src/feature/followUnfollow/FollowUnfollowComponent.js]: done. Thanks

File: src/feature/followUnfollow/__tests__/followUnfollowComponent.test.js (hunk 9 above)
Comment (raymond42) [reply src/feature/followUnfollow/__tests__/followUnfollowComponent.test.js]: Corrected

File: src/feature/followUnfollow/__tests__/followUnfollowComponent.test.js (hunk 10 above)
Comment (raymond42) [reply src/feature/followUnfollow/__tests__/followUnfollowComponent.test.js:46]: done

File: src/feature/followUnfollow/__tests__/followUnfollowComponent.test.js (hunk 11 above)
Comment (raymond42) [reply src/feature/followUnfollow/__tests__/followUnfollowComponent.test.js]: Corrected. Thanks

File: src/feature/followUnfollow/followUnfollowAction.js (hunk 13 above)
Comment (raymond42) [reply src/feature/followUnfollow/followUnfollowAction.js]: done

File: src/feature/followUnfollow/__tests__/followUnfollowComponent.test.js (hunk 12 above)
Comment (raymond42) [reply src/feature/followUnfollow/__tests__/followUnfollowComponent.test.js:71]: One is when the user is authenticated and another one is when the user is not authenticated.
It was not descriptive enough, I corrected it. Thanks

File: src/feature/articles/constants.js (hunk 3 above)
Comment (raymond42) [reply src/feature/articles/constants.js]: @bdushimi this was because I was working under the feature getting a single article which has been combined with like and dislike article so there were a lot of bugs with getting single article which was resolved on it's feature and made me pull again to get the real working codes, which is why you see some changes on like dislike article and getting single article files.
